banks-server/src/banks_server.rs

```diff
@@ -85,8 +85,21 @@ impl BanksServer {
                 .into_iter()
                 .map(|info| deserialize(&info.wire_transaction).unwrap())
                 .collect();
-            let bank = bank_forks.read().unwrap().working_bank();
-            let _ = bank.try_process_transactions(transactions.iter());
+            'inner: loop {
+                let bank = bank_forks.read().unwrap().working_bank();
+                // bank forks lock released, now verify bank hasn't been frozen yet
+                // in the mean-time the bank can not be frozen until this tx batch
+                // has been processed, this replicates BankingStage behaviour
+                let lock = bank.freeze_lock();
+                if *lock != Hash::default() {
+                    // retry reading working bank if bank is already frozen
+                    continue;
+                } else {
+                    let _ = bank.try_process_transactions(transactions.iter());
+                    // break out of inner loop and release bank freeze lock
+                    break 'inner;
+                }
```

Comment:
nit: seems more readable if an implicit `continue` is used
```suggestion
                if *lock == Hash::default() {
                    let _ = bank.try_process_transactions(transactions.iter());
                    // break out of inner loop and release bank freeze lock
                    break;
                }
                // retry reading working bank if bank is already frozen
```